### ISSUE TITLE
Subscribe to the same Topic from multiple routes in the same Application

### ISSUE THREAD

**Issue by fbridger** [dapr/dapr#3664]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Describe the feature

Currently, the Dapr SideCar does not allow to subscribe from multiple routes to the same Topic.

I was under the assumption that this basic scenario was obviously supported but I recently found out that it is not. 
Seems odd that we can't subscribe multiple/different endpoints to the same Topic from the same Application.

## Scenario to support
1. The App with id `"awesome-dapr-app"` would like to be able to subscribe to the **same Topic** (`"B"`) **from different routes** to do different tasks when that event is published.
2. "awesome-dapr-app" would return the following JSON in `GET /dapr/subscribe`
```json
[
  {
      "pubsubname": "pubsub",
      "topic": "B",
      "route": "B1"
  },
  {
      "pubsubname": "pubsub",
      "topic": "B",
      "route": "B2"
  }
]
```
3. When an event is published to the **Topic** `"B"` I would expect the Dapr SideCar to deliver the event to the routes `"/B1"` and `"/B2"` independently/concurrently/simultaneously

**Comment by yaron2** [comment]
@pkedy is that by any chance supported now in 1.4?

**Comment by pkedy** [comment]
I think what @fbridger is looking for is multiple event handlers to be invoked for a given message/event. For better or worse, that is not how the new PubSub routing feature works.  It uses rules to select 0-or-1 handler in code. Per the thread in the dotnet sdk, that was the silent behavior prior to this feature as well.

The concern I have with invoking multiple controller methods in the application is that it makes error handling difficult to reason about. For example: If there are multiple handlers, what happens when 1 fails and the others commits database transactions? Would the developer care about atomicity? Does the message get acknowledged in a partial failure scenario? If so, would the developer know to code for idempotency for each handler? At least for me, it feels much safer (and simpler) to have the handler call the series of steps a message should go through.

That said, @fbridger, I am interested in concrete scenarios where this would be useful and thoughts on the concerns above. I'm not sure what pubsub component you are using, but with Kafka, I've seen this handled simply by using multiple consumer IDs so that the handlers are completely isolated from each other. That might work for you. Feel free to reach out on Discord if you want to discuss your exact use case in more detail.

**Comment by fbridger** [comment]
@pkedy When developing event driven applications, many times you want an event to be processed by multiple consumers. I believe this is currently fully supported by Dapr but you need to consume this event (subscribe to the Topic) from different applications. 

Using diagrams from Dapr Docs

## Supported
![image](https://user-images.githubusercontent.com/6483000/133358255-e40c7983-422c-40b7-aad6-0cb713c9cdaa.png)

## Not Supported
It is not possible to do the same but from the same application
![image](https://user-images.githubusercontent.com/6483000/133358768-69edb132-e3dd-4bbe-a1a0-f7f528ee4e09.png)

_"If there are multiple handlers, what happens when 1 fails and the others commits database transactions? Would the developer care about atomicity? Does the message get acknowledged in a partial failure scenario? If so, would the developer know to code for idempotency for each handler?"_
- I would expect each handler to be independent and to be completely isolated between each other. 
- In a nutshell, I would expect to have the same behaviour as we currently have across different applications (SideCars).

In my experience, when we are developing event-driven applications this is a very common pattern to have. I just don't fully understand why we are limited to subscribe to a Topic (event) just once per application.

Currently, we are using Redis Streams as our PubSub component. 

**BTW**: It feels that **this limitation should be explicitly described in the Dapr Documentation** as this was clearly something that I would have thought it was natively supported.

I'll try to contact you in Discord. Thanks for replying guys! I really appreciate all the awesome work you have done with Dapr!

**Comment by jjcollinge** [comment]
I may not fully understand the issue/use case here so please forgive my ignorant questions. Does Dapr _actually_ need to invoke both routes on your app or can you simply have a single route that invoke both bits of functionality? The thing to consider here is that currently the Dapr sidecar will only receive a single message from the broker and then would have to invoke 2 routes on your app - so it cannot easily handle them as isolated and would need to know how to deal with partial failure. Then there's the fact that in cases where the pubsub component is using peek lock it is actually not even possible to have the same message processed in parallel (even by separate apps) as each message is locked to a consumer instance. I think this is a scenario where the semantic difference between a message and an event is important and Dapr should probably do more to highlight this.

**Comment by fbridger** [comment]
Obviously, the app could internally invoke both bits of functionality. However, by doing this, we are loosing all the benefits that pubsub provides, like the ability to distribute those 2 independent messages (or even more) between the different instances of the application or, most important, retry logic in case of unexpected failure.

To get around this, we are currently receiving that event `B` in the application and publishing 2 new events `B1` and `B2` which are also events that the same application is subscribed to.

We are basically **emulating a fan-out pattern** internally.

**Question**: Can you guys suggest a more elegant way to do this?

I fully understand the technical challenges that this feature may require and don’t expect to have it any time soon. It’s just something that I thought it was supported but I realized (after a lot of investigation and fuzz) that it was not.

To avoid other people to hit the same wall we should explicitly describe this in the documentation and suggest other ways of accomplishing this.

**Comment by pkedy** [comment]
@fbridger, You get bonus points for drawing pictures!

> I would expect each handler to be independent and to be completely isolated between each other.

In the scenario where the handlers are completely insulated from each other, could you not create a different pubsub config with a different consumerID? The message would be delivered to each handler and each would have their own independent acknowledgment/offset marking. It's really the same configuration as if it were a separate application. If this works for you, no need to read on :)

If several handlers must be called from a single topic subscription, it starts getting into workflow/DAG territory (e.g. call A then B with error handling policies). If that is the case, configuration would look like something else entirely. Also, with the possibility of partial errors, it is not completely isolated. These are some reasons why I suggest perhaps handling simple "orchestration" in the application. I'm just thinking out loud. I can totally understand the desire for this.

**Comment by fbridger** [comment]
@pkedy 
_" could you not create a different pubsub config with a different consumerID?"_
Are you thinking something like this:
 
File: `redis-pubsub-appA-group1.yaml`
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: redis-pubsub-appA-group1
  namespace: default
spec:
  type: pubsub.redis
  version: v1
  metadata:
  - name: consumerID
    value: "group1"
```

File: `redis-pubsub-appA-group2.yaml`
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: redis-pubsub-appA-group2
  namespace: default
spec:
  type: pubsub.redis
  version: v1
  metadata:
  - name: consumerID
    value: "group2"
```

Application would then exponse in GET /dapr/subcribe 
```yaml
[
  {
      "pubsubname": "redis-pubsub-appA-group1",
      "topic": "B",
      "route": "B1"
  },
  {
      "pubsubname": "redis-pubsub-appA-group2",
      "topic": "B",
      "route": "B2"
  }
]
```

**Comment by pkedy** [comment]
@fbridger Yes, precisely!

Only detail I'd add is to be know that you can also limit which apps use a component configuration using `scopes`.

https://docs.dapr.io/operations/components/component-scopes/#application-access-to-components-with-scopes

**Comment by georgestevens99** [comment]
Here is the same comment I posted to the bug .NDT-SDK #752.  It may be relevant to your work on this.

FYI A long, long time ago, before 1.0.0 was released, I did a test using the Dapr .NET SDK to verify that pubsub using both redis and Azure Service Bus Topics did indeed support multiple subscribers. I clearly remember this because I had to add a 3rd service (ServiceC) to my 2 service (ServiceA and ServiceB) test system in order to test multiple subscribers. ServiceA is the publisher.

Also keep in mind that **the essence of a Topic (aka PubSub) is its ability to publish a single event to multiple subscribers**. This is **in contrast to a Queue which can publish a message to one, and only one, listener**. So Dapr definitely needs to repair this to support the essence of a PubSub.

Thanks for your collective work.

**Comment by fbridger** [comment]
@georgestevens99 Dapr does supports multiple apps to subscribe to the same event. 

The limitation (as described in the diagram) is allowing the same app to subscribe from multiple routes to the same event.

@pkedy has suggested a workaround for this but it does require one additional pubsub config yaml to be created for each route that wants to receive that event.

Hopes this clear things a little.

**Comment by georgestevens99** [comment]
@fbridger Thanks.  From quickly reading the comments on your bug I was fearful that the ability to support multiple subscribing apps had disappeared, so I responded.  Glad to hear that!

**Comment by onionhammer** [comment]
This should be re-opened, as it was closed as stale with only a kind of hacky workaround

**Comment by danadesrosiers** [comment]
This is a basic feature that should work without additional configuration.  If I follow correctly, the consumerID is currently the app name, so multiple subscribers within the same app end up with the same subscription name and one gets ignored.  It seems like the easy fix would be to use information about the endpoint to construct the consumerID so each subscription is unique.  The app name + the url path seems like a good choice.  Actually, that's how I intuitively assumed it would work.

**Comment by siuming9512** [comment]
I also face the same problem. This is a common use case for pub/sub. Currently in Dapr, it is necessary to separate into multiple different applications to achieve this, which makes development very difficult.
Here is an example: [https://github.com/dapr/dapr/issues/2755#issuecomment-886861240](https://github.com/dapr/dapr/issues/2755#issuecomment-886861240)
We need to create 2 applications to implement this, which is very troublesome.

**Comment by onionhammer** [comment]
Bump

**Comment by thedmi** [comment]
@yaron2 Please reopen. This is still an issue.

Many apps require this functionality. Subscribing to a topic multiple times is not something exotic, it is common when messages from a single topic trigger multiple *independent* use cases, e.g. "project created" messages must notify some users AND update an external system to link the created project.

**Comment by ajstewart** [comment]
Would be great if this was possible as I ran into this today.

Had to go down the multiple pubsub component route to get the pattern I wanted. Again I wished to subscribe to a particular topic multiple times (using Service Bus Topic pubsub).

**Comment by thedmi** [comment]
So not stale

**Comment by thedmi** [comment]
Still not stale

**Comment by foliv57** [comment]
Hi, I'm facing the same problem. This scenario is very common (e.g. Having one subscription to create a read model projection + Another subscription to perform an action)).

As of now, as an example, we used multiple Azure Service Bus Topic Subscriptions with different `ServiceBusTrigger` in the same Azure Function App (Each trigger targeting a different subscription).

This pattern is powerfull because it allows to add new beaviour without having to touch a single line of existing code. (Open-Closed principle).

We are now using Dapr and are stuck on that topic. Obliged to multiply application instances just to have different `app-id` (ConsumerID). We tried multiple approaches that all looks more `workarounds` than solutions. (e.g. we also tried the "handle B => publish B1 + publish B2" mentioned here, but this means that `B` handler must be modified every time that a new behaviour must be added).

Is it not possible to have a simple solution that reuses existing logic by just allowing to provide a "consumerID" override with topic subscription metadata? So that a subscription can be setup with `pubSubName="mycomponent", topic="mytopic", consumerID="as-consumer-01"` AND `pubSubName="mycomponent", topic="mytopic", consumerID="as-consumer-02"`

Or reuse the [namespace](https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-namespace/) logic but by allowing to override the namespace for a specific subscription with matadata. e.g.: `pubSubName="mycomponent", topic="mytopic", namespace="A"` AND `pubSubName="mycomponent", topic="mytopic", namespace="B"` to have consumer groups "A.my-app-id" and "B.my-ap-id".

**Comment by foliv57** [comment]
Not stale.
